### PR TITLE
Flutter test run line markers (#1174).

### PR DESCRIPTION
**Pull request by pq**
Basic support for running tests from editor line markers.

image

In general, we should add some better scoping to our test discovery (at the moment we're not picky about the provenance of functions called `test`); TODO added and I'll follow-up with a fix, as well as some noodling on support, in general, for `testWidgets`.

Fixes: #1174.

@devoncarew

### CODE REVIEW

**Comment by devoncarew** [comment]
Wow, that was fast!

A few small comments; really excited to see this :)

File: src/io/flutter/run/test/FlutterTestLineMarkerContributor.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+public class FlutterTestLineMarkerContributor extends RunLineMarkerContributor {
+  @Nullable
+  @Override
+  public Info getInfo(@NotNull PsiElement element) {
+    if (isTestCall(element)) {
+      final AnAction[] actions = ExecutorAction.getActions();
+      final Function<PsiElement, String> tooltipProvider =
+        psiElement -> StringUtil.join(ContainerUtil.mapNotNull(actions, action -> getText(action, element)), "\n");
+      return new Info(AllIcons.RunConfigurations.TestState.Run, tooltipProvider, actions);
```
Comment (devoncarew) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: Looks like AllIcons.TestState.Green2

Comment (pq) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: Any idea what that signifies?

Here's what I see for a JUnit test:

image

Comment (pq) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: Or, I should say, what test are you using as an example?  My screenshot is from `DartSyntaxTest`...

File: src/io/flutter/run/test/FlutterTestLineMarkerContributor.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    return null;
+  }
+
+  private static boolean isTestCall(PsiElement element) {
```
Comment (pq) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: Yep.  Coming right up!

File: src/io/flutter/run/test/FlutterTestLineMarkerContributor.java (hunk 1 above)
Comment (devoncarew) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: Yeah, from the same file, I see:

screen shot 2017-07-06 at 11 44 43 am

(the editor ruler annotation, before you click on it to open the action menu)

**Comment by devoncarew** [comment]
(@jwren, @alexander-doroshko - we're adding run line markers for `flutter test` support in the flutter plugin - likely also useful for the dart plugin as well)

File: src/io/flutter/run/test/FlutterTestLineMarkerContributor.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+
+public class FlutterTestLineMarkerContributor extends RunLineMarkerContributor {
+  @Nullable
+  @Override
+  public Info getInfo(@NotNull PsiElement element) {
+    if (isTestCall(element)) {
```
Comment (pq) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: Done. 👍

**Comment by pq** [comment]
Some more progress.

  * syntax tests (we'll want integration ones to test scoping)
  * scoping to contents in `test/`


PTAL.

File: src/io/flutter/run/test/FlutterTestLineMarkerContributor.java (hunk 1 above)
Comment (pq) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: FWIW: I have a bead on what's up here.  Will follow-up in another PR.

Comment (devoncarew) [inline src/io/flutter/run/test/FlutterTestLineMarkerContributor.java]: 👍

File: src/io/flutter/run/FlutterRunConfigurationProducer.java [hunk 4]
```diff
@@ -84,6 +84,14 @@ public boolean shouldReplace(@NotNull ConfigurationFromContext self, @NotNull Co
...
   public static VirtualFile getFlutterEntryFile(final @NotNull ConfigurationContext context, boolean requireFlutterImport, boolean omitTests) {
     final DartFile dart = getDartFile(context);
+    return getFlutterEntryFile(dart, requireFlutterImport, omitTests);
+  }
+
+  /**
+   * Returns the corresponding virtual file containing a Flutter app's main() function, or null if not a match.
+   */
+  @Nullable
+  public static VirtualFile getFlutterEntryFile(final @Nullable DartFile dart , boolean requireFlutterImport, boolean omitTests) {
```
Comment (devoncarew) [inline src/io/flutter/run/FlutterRunConfigurationProducer.java]: nit re: the formatting here (space before the comma)

**Comment by devoncarew** [comment]
lgtm!

File: src/io/flutter/run/FlutterRunConfigurationProducer.java (hunk 4 above)
Comment (pq) [inline src/io/flutter/run/FlutterRunConfigurationProducer.java]: Good eyes! 🦅

**Pull request merged**